src/appstatus.cpp

```diff
@@ -0,0 +1,171 @@
+#include "appstatus.h"
+
+AppStatus::AppStatus(QObject *parent)
+    : QObject(parent)
+    , m_status(Disabled)
+    , m_downloadStatus()
+    , m_uploadStatus()
+    , m_updating()
+    , m_bluetoothBusy()
+    , m_bluetoothEnabled()
+    , m_atRisk()
+    , m_downloadAvailable()
+{
+}
+
+AppStatus::Status AppStatus::status() const
+{
+    return m_status;
+}
+
+void AppStatus::updateStatus()
+{
+    Status status = m_status;
+
+    if (m_updating) {
+        m_status = Updating;
+    }
+    else if ((m_uploadStatus != Upload::StatusIdle) && (m_uploadStatus != Upload::StatusError)) {
+        m_status = Uploading;
+    }
+    else if ((m_downloadStatus != Download::StatusIdle) && (m_downloadStatus != Download::StatusError)) {
+        m_status = Downloading;
+    }
+    else if (m_uploadStatus == Upload::StatusError) {
+        m_status = ErrorUploading;
+    }
+    else if (m_downloadStatus == Download::StatusError) {
+        m_status = ErrorDownloading;
+    }
+    else if (m_bluetoothBusy) {
+        m_status = BluetoothBusy;
+    }
+    else if (m_atRisk) {
+        m_status = AtRisk;
+    }
+    else if (m_downloadAvailable) {
+        m_status = DailyUpdateRequired;
+    }
+    else if (m_bluetoothEnabled) {
+        m_status = Active;
+    }
+    else {
+        m_status = Disabled;
+    }
+
+    if (m_status != status) {
+        emit statusChanged();
+    }
+}
+
+Download::Status AppStatus::downloadStatus() const
+{
+    return m_downloadStatus;
+}
+
+Upload::Status AppStatus::uploadStatus() const
+{
+    return m_uploadStatus;
+}
+
+bool AppStatus::updating() const
+{
+    return m_updating;
+}
+
+bool AppStatus::bluetoothBusy() const
+{
+    return m_bluetoothBusy;
+}
+
+bool AppStatus::bluetoothEnabled() const
+{
+    return m_bluetoothEnabled;
+}
+
+bool AppStatus::atRisk() const
+{
+    return m_atRisk;
+}
+
+bool AppStatus::downloadAvailable() const
+{
+    return m_downloadAvailable;
+}
+
+void AppStatus::setDownloadStatus(Download::Status downloadStatus)
+{
+    if (m_downloadStatus != downloadStatus) {
+        m_downloadStatus = downloadStatus;
+        emit downloadStatusChanged();
+        updateStatus();
```

Comment:
When `downloadStatusChanged()` already was emitted, could't be `updateStatus()` as slot connected to it?

Reply:
It could be, but in practice the result would be identical, but with just a bit of extra overhead. When the code is all inside the same class and same thread, it's better to call the method directly (unless there's something I'm missing here?).